libdnf/utils/utils.cpp

```diff
@@ -184,6 +191,48 @@ std::vector<std::string> getDirContent(const std::string &dirPath)
     return content;
 }
 
+void decompress(const char * inPath, const char * outPath, mode_t outMode, const char * compressType)
+{
+    auto inFd = open(inPath, O_RDONLY);
+    if (inFd == -1)
+        throw std::runtime_error(tfm::format("open: %s", strerror(errno)));
+    if (!compressType)
+        compressType = inPath;
+    auto inFile = solv_xfopen_fd(compressType, inFd, "r");
+    if (inFile == NULL) {
+        close(inFd);
+        throw std::runtime_error("solv_xfopen_fd: Can't open stream");
+    }
+    auto outFd = open(outPath, O_WRONLY | O_CREAT | O_TRUNC, outMode);
+    if (outFd == -1) {
+        int err = errno;
+        fclose(inFile);
+        throw std::runtime_error(tfm::format("open: %s", strerror(err)));
+    }
+    char buf[4096];
+    while (auto readBytes = fread(buf, 1, sizeof(buf), inFile)) {
+        auto writtenBytes = write(outFd, buf, readBytes);
+        if (writtenBytes == -1) {
+            int err = errno;
+            close(outFd);
+            fclose(inFile);
+            throw std::runtime_error(tfm::format("write: %s", strerror(err)));
+        }
+        if (writtenBytes != static_cast<int>(readBytes)) {
```

Comment:
It is not an error for write() to write less bytes than it was asked to, it can happen due to various reasons. See the man page.